File: programmers/kakao#2/python/KAKAO7.py
import collections
def solution(record):
    
    ENTER_FORMAT = "%s님이 들어왔습니다."
    LEAVE_FORMAT = "%s님이 나갔습니다."
    
    format_dict = {"Enter": ENTER_FORMAT, "Leave": LEAVE_FORMAT}
    
    user_infos = collections.defaultdict(str)
    command_list = []
    
    for log in record:
        log_split = log.split(" ")
        
        if log_split[0] == "Enter":
            user_infos[log_split[1]] = log_split[2] 
            command_list.append([log_split[0], log_split[1]])
            
        elif log_split[0] == "Leave":
            command_list.append([log_split[0], log_split[1]])
            
        else:
            user_infos[log_split[1]] = log_split[2]
            
    return [format_dict[cmd[0]] % format(user_infos[cmd[1]]) for cmd in command_list]

"""
[카카오 2019 공채] 오픈채팅방
해결: O
시간: 16분
"""

File: 0_4.programmers/kakao#2/java/KAKAO7.java
import java.util.*;
public class KAKAO7 {
    class Solution {
        private static final String ENTER_FORMAT = "%s님이 들어왔습니다.";
        private static final String LEAVE_FORMAT = "%s님이 나갔습니다.";
        
        private Map<String, User> userInfo = new HashMap<String, User>();
        
        private static class User {
            private String id;
            private String nickname;
            
            User(String id, String nickname) {
                this.id = id;
                this.nickname = nickname;
            }
        }
        
        private static class Command {
            private String command;
            private String userId;
            
            Command(String command, String userId) {
                this.command = command;
                this.userId = userId;
            }
        }
        
        public String[] solution(String[] record) {
            List<Command> commandList = new ArrayList<>();
            
            for (String log : record) {
                String[] logSplit = log.split(" ");
                
                if (logSplit[0].equals("Enter")) {
                    userInfo.put(logSplit[1], new User(logSplit[1], logSplit[2]));
                    commandList.add(new Command(logSplit[0], logSplit[1]));
                } else if (logSplit[0].equals("Change")) {
                    userInfo.get(logSplit[1]).nickname = logSplit[2];
                } else {
                    commandList.add(new Command(logSplit[0], logSplit[1]));
                }
            }
            
            
            return commandList.stream()
                .map(cmd -> String.format(cmd.command.equals("Enter") ? ENTER_FORMAT : LEAVE_FORMAT, userInfo.get(cmd.userId).nickname))
                .toArray(ary -> new String[commandList.size()]);
        }
    }
}
/***
 * [카카오 2019 공채] 오픈 채팅방
 * 해결: O
 * 시간: 16분
 * 자바는 객체!객체!객체!!!!! 
 */